vm/src/org/mozartoz/truffle/translator/Translator.java

```diff
@@ -331,14 +340,20 @@ public OzNode translateProc(ProcExpression procExpression) {
 					throw unknown("variable", variable);
 				}
 			}
-
 			nodes[i] = translate(procExpression.body());
 
 			OzNode procBody = SequenceNode.sequence(nodes);
+			if (this.isSelfTailRec) { // Side-affected by translate of some SelfTailCallMarker
+				if (Options.SELF_TAIL_CALLS_OSR)
+					procBody = new SelfTailCallCatcherNode(procBody);
+				else
+					procBody = new SelfTailCallCatcherNode.NoOSR(procBody);
```

Comment:
I would hide this in a factory method like `SelfTailCallCatcherNode.create(body)`.